internal/worker/objectstoredrainer/worker.go

```diff
@@ -5,66 +5,199 @@ package objectstoredrainer
+	ObjectStoreType           objectstore.BackendType
+	NewHashFileSystemAccessor NewHashFileSystemAccessorFunc
+	NewDrainerWorker          NewDrainerWorkerFunc
+	S3Client                  objectstore.Client
+	SelectFileHash            SelectFileHashFunc
+	RootDir                   string
+	RootBucketName            string
+	Logger                    logger.Logger
+	Clock                     clock.Clock
 }
 
 // Validate returns an error if the config cannot be expected to
 // drive a functional Worker.
 func (config Config) Validate() error {
+	if config.Agent == nil {
+		return errors.Errorf("nil Agent").Add(coreerrors.NotValid)
+	}
 	if config.Guard == nil {
-		return errors.NotValidf("nil Guard")
+		return errors.Errorf("nil Guard").Add(coreerrors.NotValid)
+	}
+	if config.GuardService == nil {
+		return errors.Errorf("nil GuardService").Add(coreerrors.NotValid)
+	}
+	if config.ControllerService == nil {
+		return errors.Errorf("nil ControllerService").Add(coreerrors.NotValid)
 	}
-	if config.ObjectStoreService == nil {
-		return errors.NotValidf("nil ObjectStoreService")
+	if config.ControllerConfigService == nil {
+		return errors.Errorf("nil ControllerConfigService").Add(coreerrors.NotValid)
+	}
+	if config.ObjectStoreServicesGetter == nil {
+		return errors.Errorf("nil ObjectStoreServicesGetter").Add(coreerrors.NotValid)
+	}
+	if config.ObjectStoreFlusher == nil {
+		return errors.Errorf("nil ObjectStoreFlusher").Add(coreerrors.NotValid)
+	}
+	if config.NewHashFileSystemAccessor == nil {
+		return errors.Errorf("nil NewHashFileSystemAccessor").Add(coreerrors.NotValid)
+	}
+	if config.NewDrainerWorker == nil {
+		return errors.Errorf("nil NewDrainerWorker").Add(coreerrors.NotValid)
+	}
+	if config.S3Client == nil {
+		return errors.Errorf("nil S3Client").Add(coreerrors.NotValid)
+	}
+	if config.SelectFileHash == nil {
+		return errors.Errorf("nil SelectFileHash").Add(coreerrors.NotValid)
+	}
+	if config.RootDir == "" {
+		return errors.Errorf("empty RootDir").Add(coreerrors.NotValid)
+	}
+	if config.RootBucketName == "" {
+		return errors.Errorf("empty RootBucketName").Add(coreerrors.NotValid)
 	}
 	if config.Logger == nil {
-		return errors.NotValidf("nil Logger")
+		return errors.Errorf("nil Logger").Add(coreerrors.NotValid)
+	}
+	if config.Clock == nil {
+		return errors.Errorf("nil Clock").Add(coreerrors.NotValid)
 	}
 	return nil
 }
 
 // NewWorker returns a Worker that tracks the result of the configured.
-func NewWorker(ctx context.Context, config Config) (worker.Worker, error) {
+func NewWorker(config Config) (worker.Worker, error) {
 	if err := config.Validate(); err != nil {
-		return nil, errors.Trace(err)
+		return nil, errors.Capture(err)
+	}
+
+	runner, err := worker.NewRunner(worker.RunnerParams{
+		Name: "objectstore-drainer",
+		IsFatal: func(err error) bool {
+			return false
+		},
+		ShouldRestart: func(err error) bool {
+			return true
+		},
+		RestartDelay: time.Second * 10,
```

Comment:
Quite long isn't it?

Reply:
Yeah, probably, I really want to get https://github.com/juju/worker/pull/41 in at some point, which gives us the best of both worlds.

Reply:
All our workers restart after 10 seconds... I'm happy to revisit this as a whole.